java/servers/src/org/xtreemfs/mrc/quota/MRCVoucherManager.java

```diff
@@ -0,0 +1,638 @@
+/*
+ * Copyright (c) 2015 by Robert Bärhold, Zuse Institute Berlin
+ *
+ * Licensed under the BSD License, see LICENSE file for details.
+ *
+ */
+
+package org.xtreemfs.mrc.quota;
+
+import java.util.Set;
+
+import org.xtreemfs.common.quota.QuotaConstants;
+import org.xtreemfs.foundation.logging.Logging;
+import org.xtreemfs.foundation.pbrpc.generatedinterfaces.RPC.POSIXErrno;
+import org.xtreemfs.mrc.MRCException;
+import org.xtreemfs.mrc.UserException;
+import org.xtreemfs.mrc.ac.FileAccessManager;
+import org.xtreemfs.mrc.database.AtomicDBUpdate;
+import org.xtreemfs.mrc.database.DatabaseException;
+import org.xtreemfs.mrc.database.DatabaseResultSet;
+import org.xtreemfs.mrc.database.StorageManager;
+import org.xtreemfs.mrc.metadata.BufferBackedFileVoucherClientInfo;
+import org.xtreemfs.mrc.metadata.BufferBackedFileVoucherInfo;
+import org.xtreemfs.mrc.metadata.FileVoucherClientInfo;
+import org.xtreemfs.mrc.metadata.FileVoucherInfo;
+
+/**
+ * This class manages all voucher requested affairs and if necessary, it delegates them to reference classes.
+ * 
+ * TODO: This class is currently thread-safe, but because the MRC has only one operational thread, there is no need for
+ * this. False: XLockCoordinator handling add/remove replica works in a separate thread.
+ */
+public class MRCVoucherManager {
+
+    private final MRCQuotaManager mrcQuotaManager;
+
+    public MRCVoucherManager(MRCQuotaManager mrcQuotaManager) {
+        this.mrcQuotaManager = mrcQuotaManager;
+    }
+
+    /**
+     * Returns the first voucher for a given file, client and expire time, if no quota is violated. If no quota is set
+     * at all, it will return an unlimited voucher.
+     * 
+     * @param quotaFileInformation
+     * @param clientId
+     * @param expireTime
+     * @param update
+     * @return
+     * @throws UserException
+     */
+    public synchronized long getVoucher(QuotaFileInformation quotaFileInformation, String clientId, long expireTime,
```

Comment:
Please check if the synchronized modifiers can be removed in this class. BabuDB is able to handle concurrent operations.